jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java

```diff
@@ -990,36 +990,42 @@ private CloseConnections(Set<Closeable> closed)
         @Override
         public void update(Selector selector)
         {
-            if (LOG.isDebugEnabled())
-                LOG.debug("Closing {} connections on {}", selector.keys().size(), ManagedSelector.this);
-            for (SelectionKey key : selector.keys())
+            try
             {
-                if (key != null && key.isValid())
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Closing {} connections on {}", selector.keys().size(), ManagedSelector.this);
+                for (SelectionKey key : selector.keys())
                 {
-                    Closeable closeable = null;
-                    Object attachment = key.attachment();
-                    if (attachment instanceof EndPoint)
+                    if (key != null && key.isValid())
                     {
-                        EndPoint endPoint = (EndPoint)attachment;
-                        Connection connection = endPoint.getConnection();
-                        closeable = Objects.requireNonNullElse(connection, endPoint);
-                    }
-
-                    if (closeable != null)
-                    {
-                        if (_closed == null)
+                        Closeable closeable = null;
+                        Object attachment = key.attachment();
+                        if (attachment instanceof EndPoint)
                         {
-                            IO.close(closeable);
+                            EndPoint endPoint = (EndPoint)attachment;
+                            Connection connection = endPoint.getConnection();
+                            closeable = Objects.requireNonNullElse(connection, endPoint);
                         }
-                        else if (!_closed.contains(closeable))
+
+                        if (closeable != null)
                         {
-                            _closed.add(closeable);
-                            IO.close(closeable);
+                            if (_closed == null)
+                            {
+                                IO.close(closeable);
+                            }
+                            else if (!_closed.contains(closeable))
+                            {
+                                _closed.add(closeable);
+                                IO.close(closeable);
+                            }
                         }
                     }
                 }
             }
-            _complete.countDown();
+            finally
```

Comment:
Is this fixing a known bug or just defensive in case something above throws?
I can't see what could throw above???

Reply:
`IO.close()` can throw.

Reply:
`IO.close(Closeable)` consumes `IOException` now, specifically to not throw.
Are you saying some other exception is flowing out?

Reply:
Yes. and if it does, without the `finally` block, we wait forever.